// sunlight-x/babel-out/languages/brainfuck.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.numberParser = numberParser;
// sunlight-x: Intelligent Syntax Highlighting, Modernized
// Copyright 2017 Leung Wing-chung. All rights reserved.
// Use of this source code is governed by a Apache License Version 2.0, that can
// be found in the LICENSE file.

var name = exports.name = "brainfuck";

var customTokens = exports.customTokens = {
  increment: {
    values: [">"],
    boundary: ""
  },
  decrement: {
    values: ["<"],
    boundary: ""
  },
  incrementPointer: {
    values: ["+"],
    boundary: ""
  },
  decrementPointer: {
    values: ["-"],
    boundary: ""
  },
  write: {
    values: ["."],
    boundary: ""
  },
  read: {
    values: [","],
    boundary: ""
  },
  openLoop: {
    values: ["["],
    boundary: ""
  },
  closeLoop: {
    values: ["]"],
    boundary: ""
  }
};

var punctuation = exports.punctuation = /(?!x)x/;

/**
 * Empty number parser. Always rejects the token.
 * @returns {null}
 */
function numberParser() {
  return null;
}
