// sunlight-x/babel-out/languages/javascript.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.operators = exports.namedIdentRules = exports.identAfterFirstLetter = exports.identFirstLetter = exports.customParseRules = exports.scopes = exports.customTokens = exports.keywords = exports.name = undefined;

var _util = require("../util.js");

var util = _interopRequireWildcard(_util);

var _regexp = require("./common/regexp.js");

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

var name = exports.name = "javascript";

var keywords = exports.keywords = [
// keywords
"break", "case", "catch", "continue", "default", "delete", "do", "else", "finally", "for", "function", "if", "in", "instanceof", "new", "return", "switch", "this", "throw", "try", "typeof", "var", "void", "while", "with",

// literals
"true", "false", "null"];

var customTokens = exports.customTokens = {
  reservedWord: {
    values: ["abstract", "boolean", "byte", "char", "class", "const", "debugger", "double", "enum", "export const", "extends", "final", "float", "goto", "implements", "import", "int", "interface", "long", "native", "package", "private", "protected", "public", "short", "static", "super", "synchronized", "throws", "transient", "volatile"],
    boundary: "\\b"
  },

  globalVariable: {
    values: ["NaN", "Infinity", "undefined"],
    boundary: "\\b"
  },

  globalFunction: {
    values: ["encodeURI", "encodeURIComponent", "decodeURI", "decodeURIComponent", "parseInt", "parseFloat", "isNaN", "isFinite", "eval"],
    boundary: "\\b"
  },

  globalObject: {
    values: ["Math", "JSON", "XMLHttpRequest", "XDomainRequest", "ActiveXObject", "Boolean", "Date", "Array", "Image", "Function", "Object", "Number", "RegExp", "String"],
    boundary: "\\b"
  }
};

var scopes = exports.scopes = {
  string: [['"', '"', util.escapeSequences.concat(['\\"']), false], ["'", "'", util.escapeSequences.concat(["\\'", "\\\\"]), false]],
  comment: [["//", "\n", [], true], ["/*", "*/", [], false]]
};

var customParseRules = exports.customParseRules = [_regexp.ParseRegExpLiteral];

var identFirstLetter = exports.identFirstLetter = /[$A-Za-z_]/;
var identAfterFirstLetter = exports.identAfterFirstLetter = /[\w$]/;

var namedIdentRules = exports.namedIdentRules = {
  follows: [[{ token: "keyword", values: ["function"] }, util.whitespace]]
};

var operators = exports.operators = [
// arithmetic
"++", "+=", "+", "--", "-=", "-", "*=", "*", "/=", "/", "%=", "%",

// boolean
"&&", "||",

// bitwise
"|=", "|", "&=", "&", "^=", "^", ">>>=", ">>>", ">>=", ">>", "<<=", "<<",

// inequality
"<=", "<", ">=", ">", "===", "==", "!==", "!=",

// unary
"!", "~",

// other
"?", ":", ".", "="];
