// sunlight-x/babel-out/languages/sln.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.operators = exports.namedIdentRules = exports.identAfterFirstLetter = exports.identFirstLetter = exports.scopes = exports.customTokens = exports.keywords = exports.name = undefined;

var _util = require("../util.js");

var util = _interopRequireWildcard(_util);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

var name = exports.name = "sln"; // sunlight-x: Intelligent Syntax Highlighting, Modernized
// Copyright 2017 Leung Wing-chung. All rights reserved.
// Use of this source code is governed by a Apache License Version 2.0, that can
// be found in the LICENSE file.

var keywords = exports.keywords = ["Project", "EndProject", "GlobalSection", "Global", "EndGlobalSection", "EndGlobal"];

var customTokens = exports.customTokens = {
  sectionName: {
    values: ["preSolution", "postSolution"],
    boundary: "\\b"
  }
};

var scopes = exports.scopes = {
  string: [['"', '"', ['\\"', "\\\\"], false]],
  comment: [["#", "\n", [], true]],
  guid: [["{", "}", [], false]],
  formatDeclaration: [["Microsoft Visual Studio Solution File, Format Version", "\n", [], true]]
};

var identFirstLetter = exports.identFirstLetter = /[A-Za-z_.]/;
var identAfterFirstLetter = exports.identAfterFirstLetter = /[\w-.]/;

var namedIdentRules = exports.namedIdentRules = {
  follows: [[{ token: "keyword", values: ["GlobalSection"] }, util.whitespace, { token: "punctuation", values: ["("] }, util.whitespace]]
};

var operators = exports.operators = ["=", "|"];
